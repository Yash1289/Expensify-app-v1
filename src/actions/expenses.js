import { v4 as uuidv4 } from "uuid"

export const expenseAdder = ({ description = "", note = "", amount = 0, createdAt = undefined } = {}) => ({
    type: "ExpenseAdd",
    expense: { id: uuidv4(), description, note, amount, createdAt }
});

export const expenseRemover = ({ id } = {}) => ({
    type: "ExpenseRemove",
    id
})

export const expenseEdit = (id, update) => ({
    type: "ExpenseEdit",
    id,
    update
})