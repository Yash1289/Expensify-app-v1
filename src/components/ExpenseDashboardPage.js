const React = require("react");
import ConnectedExpensesList from  "./ExpensesList"
import ExpensesListFilter from  "./ExpensesListFilter"

const ExpenseDashboardPage = () => (
    <div>
        <p> this is form my dashboard component</p>
        <ConnectedExpensesList />
        <ExpensesListFilter />
    </div>
)

export default ExpenseDashboardPage ;