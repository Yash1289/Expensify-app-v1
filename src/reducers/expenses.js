

const expenseReducerDefaultState = []

const expenseReducer = (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case "ExpenseAdd":
            return [
                ...state, action.expense
            ];
        case "ExpenseRemove":
            return state.filter((expense) => expense.id !== action.id)
        case "ExpenseEdit":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return { ...expense, ...action.update }
                }
                else {
                    return expense;
                }
            })
        default:
            return state;
    }
}

export default expenseReducer