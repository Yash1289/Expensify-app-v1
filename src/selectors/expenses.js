import moment from "moment"

const filterExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    const filteredExpenses = expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true ;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true ; 
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch;
    })

    return filteredExpenses.sort((a, b) => {
        if (sortBy === "date") {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        if (sortBy === "amount") {
            return a.amount < b.amount ? 1 : -1
        }
    })
}

export default filterExpenses