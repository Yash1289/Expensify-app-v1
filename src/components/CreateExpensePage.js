const React = require("react");
import ExpenseForm from "./ExpenseForm"
import { expenseAdder } from "../actions/expenses"
import { connect } from "react-redux"

export class CreateExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.CreateExpense(expense)
        this.props.history.push("/")
    }

    render(){
        return (
            <div>
                <p>Add Expense</p>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => (
    {
        CreateExpense : (expense) => dispatch(expenseAdder(expense))
    }
)
export default connect(undefined , mapDispatchToProps)(CreateExpensePage)