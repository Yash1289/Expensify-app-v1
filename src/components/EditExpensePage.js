const React = require("react");
import { connect } from "react-redux"
import ExpenseForm from "./ExpenseForm"
import { startExpenseEdit , startExpenseRemover } from "../actions/expenses"

export class EditExpensePage extends React.Component { 
    onSubmit = (expense) => {
        this.props.ExpenseEdit(this.props.expense.id , expense)
        this.props.history.push("/")
    }
    onClick = () => {
        console.log(this.props.expense.id)
        this.props.ExpenseRemover({ id: this.props.expense.id })
        this.props.history.push("/")
    }
    render() {
        return (
            <div>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
                <button onClick={this.onClick} >Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state , props ) => {
    return {
        expense : state.expenses.find((expense) => expense.id === props.match.params.id )
    }
}

const mapDispatchToProps = ( dispatch , props ) => {
    return {
        ExpenseEdit: (id , expense) => dispatch(startExpenseEdit(id , expense)),
        ExpenseRemover : (data) => dispatch(startExpenseRemover(data))
    }
}

export default connect(mapStateToProps , mapDispatchToProps)(EditExpensePage) 