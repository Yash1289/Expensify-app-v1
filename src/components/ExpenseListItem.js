import React from "react";
import {expenseRemover} from "../actions/expenses"
import { NavLink } from "react-router-dom"
import { connect } from "react-redux";

const ExpenseListItem = ({  id , description , amount, createdAt}) => (
    <div>
        <NavLink to={`/edit/${id}`} ><h3>{description}</h3></NavLink>
        <p>{amount} - {createdAt}</p>
    </div>
)



export default ExpenseListItem ;