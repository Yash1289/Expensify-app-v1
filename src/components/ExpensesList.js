import React from 'react'
import ExpenseListItem from "./ExpenseListItem"
import filterExpenses from "../selectors/expenses"
import { connect } from "react-redux"



export const ExpensesList = (props) => (
    <div>
        {props.expenses.length === 0 && <p>No Expenses</p>}
        {
            props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: filterExpenses(state.expenses, state.filter)
    }
}

const ConnectedExpensesList = connect(mapStateToProps)(ExpensesList)

export default ConnectedExpensesList