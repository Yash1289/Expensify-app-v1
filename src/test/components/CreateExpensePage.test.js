import React from "react"
import { shallow } from "enzyme"
import { CreateExpensePage } from "../../components/CreateExpensePage"
import expenses from "../fixtures/expenses"

let CreateExpense, history, wrapper ;

beforeEach(() => {
     CreateExpense = jest.fn()
     history = {
        push: jest.fn()
    }
     wrapper = shallow(<CreateExpensePage CreateExpense={CreateExpense} history={history} />)
})

test("Should render CreateExpensePage correctly", () => {
    expect(wrapper).toMatchSnapshot()
}) 

test("Should call onSubmit with the necessary arguments", () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[2])
    expect(CreateExpense).toHaveBeenLastCalledWith(expenses[2])
    expect(history.push).toHaveBeenLastCalledWith("/")
})   