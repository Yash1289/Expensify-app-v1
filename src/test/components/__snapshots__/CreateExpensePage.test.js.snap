// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CreateExpensePage correctly 1`] = `
<div>
  <p>
    Add Expense
  </p>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
